ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp

```diff
@@ -62,73 +63,106 @@ class IsCrackTip
+    for (unsigned frac_id=0; frac_id<vec_fracture_mat_IDs.size(); frac_id++)
+    {
+        const auto frac_mat_id = vec_fracture_mat_IDs[frac_id];
+        std::vector<MeshLib::Element*> &vec_elements = vec_fracture_elements[frac_id];
+        std::copy_if(all_fracture_elements.begin(), all_fracture_elements.end(),
+                     std::back_inserter(vec_elements),
+                     [&](MeshLib::Element* e) {
+                         return (*opt_material_ids)[e->getID()] == frac_mat_id;
+                     });
+        DBUG("-> found %d elements on the fracture %d", vec_elements.size(), frac_id);
+    }
 
     // get a vector of fracture nodes
-    for (MeshLib::Element* e : vec_fracture_elements)
+    vec_fracture_nodes.resize(vec_fracture_mat_IDs.size());
+    for (unsigned frac_id=0; frac_id<vec_fracture_mat_IDs.size(); frac_id++)
     {
-        for (unsigned i=0; i<e->getNumberOfNodes(); i++)
+        std::vector<MeshLib::Node*> &vec_nodes = vec_fracture_nodes[frac_id];
+        for (MeshLib::Element* e : vec_fracture_elements[frac_id])
         {
-            if (isCrackTip(*e->getNode(i)))
-                continue;
-            vec_fracture_nodes.push_back(const_cast<MeshLib::Node*>(e->getNode(i)));
+            for (unsigned i=0; i<e->getNumberOfNodes(); i++)
+            {
+                if (isCrackTip(*e->getNode(i)))
+                    continue;
+                vec_nodes.push_back(const_cast<MeshLib::Node*>(e->getNode(i)));
+            }
         }
+        BaseLib::makeVectorUnique(
+            vec_nodes,
+            [](MeshLib::Node* node1, MeshLib::Node* node2) {
+                return node1->getID() < node2->getID();
+            });
+        DBUG("-> found %d nodes on the fracture %d", vec_nodes.size(), frac_id);
     }
-    std::sort(vec_fracture_nodes.begin(), vec_fracture_nodes.end(),
-        [](MeshLib::Node* node1, MeshLib::Node* node2) { return (node1->getID() < node2->getID()); }
-        );
-    vec_fracture_nodes.erase(
-                std::unique(vec_fracture_nodes.begin(), vec_fracture_nodes.end()),
-                vec_fracture_nodes.end());
-    DBUG("-> found %d nodes on the fracture", vec_fracture_nodes.size());
 
     // create a vector fracture elements and connected matrix elements,
     // which are passed to a DoF table
-    // first, collect matrix elements
-    for (MeshLib::Element *e : vec_fracture_elements)
+    for (auto fracture_elements : vec_fracture_elements)
     {
-        for (unsigned i=0; i<e->getNumberOfBaseNodes(); i++)
+        std::vector<MeshLib::Element*> vec_ele;
+        // first, collect matrix elements
+        for (MeshLib::Element*e : fracture_elements)
         {
-            MeshLib::Node const* node = e->getNode(i);
-            if (isCrackTip(*node))
-                continue;
-            for (unsigned j=0; j<node->getNumberOfElements(); j++)
+            for (unsigned i=0; i<e->getNumberOfBaseNodes(); i++)
```

Comment:
Maybe add a comment that it is sufficient to iterate over base nodes, because they are already connected to all neighbours?